From 6b022e2fa68c501d8b7e804191dc4f07bf149ab4 Mon Sep 17 00:00:00 2001
From: Mohan Kumar G <mkumarg@qti.qualcomm.com>
Date: Thu, 24 Jul 2025 16:16:18 +0530
Subject: [PATCH 2/4] UPSTREAM:PROTOCOL:HOSTAPD: Update sta profiles for link 
 addition to ml_reconf_req

Update the STA profiles required for link addition request, parsed
from link_reconf_req_list, into ml_reconf_req structure. Also, add
bitmaps for all the updated links. Use this structure to hold
all reconfig request information and per sta profiles for
requested links.

Patch-work: None

Depends-on: None

Feature: ML_RECONFIG

Signed-off-by: Mohan Kumar G <mkumarg@qti.qualcomm.com>
---
 src/ap/ap_drv_ops.c     | 24 +++++++++++++++++++++++-
 src/ap/ap_drv_ops.h     |  2 +-
 src/ap/ieee802_11.c     | 14 ++++++++------
 src/ap/ieee802_11_eht.c | 10 ++++++++++
 src/ap/sta_info.c       |  2 +-
 5 files changed, 43 insertions(+), 9 deletions(-)

diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index 52d7d91..92d9a4b 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -533,9 +533,11 @@ int hostapd_sta_add(struct hostapd_data *hapd,
 		    int set, const u8 *link_addr, bool mld_link_sta,
 		    u16 eml_cap,
 		    const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab,
-		    size_t eht_240mhz_len)
+		    size_t eht_240mhz_len, int type)
 {
 	struct hostapd_sta_add_params params;
+	struct hostapd_data *assoc_hapd;
+	struct sta_info *sta, *assoc_sta;
 
 	if (hapd->driver == NULL)
 		return 0;
@@ -579,6 +581,26 @@ int hostapd_sta_add(struct hostapd_data *hapd,
 		if (link_addr)
 			params.eml_cap = eml_cap;
 	}
+
+	if (type == LINK_PARSE_RECONF) {
+		sta = ap_get_sta(hapd, addr);
+		if (!sta) {
+			wpa_printf(MSG_ERROR, "No link sta for addr = " MACSTR,
+				   MAC2STR(addr));
+			return -1;
+		}
+		assoc_sta = hostapd_ml_get_assoc_sta(hapd, sta, &assoc_hapd);
+		if (!assoc_sta) {
+			wpa_printf(MSG_ERROR, "No assoc sta found");
+			return -1;
+		}
+		assoc_sta->recfg_sta_add_params[hapd->mld_link_id] =
+			os_zalloc(sizeof(struct hostapd_sta_add_params));
+
+		os_memcpy(assoc_sta->recfg_sta_add_params[hapd->mld_link_id],
+			  &params, sizeof(params));
+		return 0;
+	}
 #endif /* CONFIG_IEEE80211BE */
 
 	return hapd->driver->sta_add(hapd->drv_priv, &params);
diff --git a/src/ap/ap_drv_ops.h b/src/ap/ap_drv_ops.h
index d738b52..d72af0d 100644
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -52,7 +52,7 @@ int hostapd_sta_add(struct hostapd_data *hapd,
 		    int set, const u8 *link_addr, bool mld_link_sta,
 		    u16 eml_cap,
 		    const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab,
-		    size_t eht_240mhz_len);
+		    size_t eht_240mhz_len, int type);
 int hostapd_set_privacy(struct hostapd_data *hapd, int enabled);
 int hostapd_set_generic_elem(struct hostapd_data *hapd, const u8 *elem,
 			     size_t elem_len);
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 081f316..7486cfa 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5104,7 +5104,7 @@ int ieee80211_ml_process_link(struct hostapd_data *hapd,
 	/* TODO: What other processing is required? */
 
 	if (!offload &&
-	    add_associated_sta(hapd, sta, type == LINK_PARSE_REASSOC))
+	    add_associated_sta(hapd, sta, type))
 		status = WLAN_STATUS_AP_UNABLE_TO_HANDLE_NEW_STA;
 out:
 	wpabuf_free(mlbuf);
@@ -5192,7 +5192,7 @@ int hostapd_process_assoc_ml_info(struct hostapd_data *hapd,
 					    0, NULL, NULL, NULL, 0, NULL, 0, NULL,
 					    sta->flags, 0, 0, 0, 0,
 					    mld_link_addr, mld_link_sta, eml_cap,
-					    NULL, 0)) {
+					    NULL, 0, reassoc)) {
 				hostapd_logger(hapd, sta->addr,HOSTAPD_MODULE_IEEE80211,HOSTAPD_LEVEL_NOTICE,
 					       "Could not add STA to kernel driver");
 				return -1;
@@ -5268,7 +5268,7 @@ static void send_deauth(struct hostapd_data *hapd, const u8 *addr,
 
 
 static int add_associated_sta(struct hostapd_data *hapd,
-			      struct sta_info *sta, int reassoc)
+			      struct sta_info *sta, int type)
 {
 	struct ieee80211_ht_capabilities ht_cap;
 	struct ieee80211_vht_capabilities vht_cap;
@@ -5279,6 +5279,7 @@ static int add_associated_sta(struct hostapd_data *hapd,
 	const u8 *mld_link_addr = NULL;
 	bool mld_link_sta = false;
 	u16 eml_cap = 0;
+	bool reassoc = (type == LINK_PARSE_REASSOC);
 
 #ifdef CONFIG_IEEE80211BE
 	if (ap_sta_is_mld(hapd, sta)) {
@@ -5320,12 +5321,12 @@ static int add_associated_sta(struct hostapd_data *hapd,
 	 * STA entry is removed from the list.
 	 */
 	wpa_printf(MSG_DEBUG, "Add associated STA " MACSTR
-		   " (added_unassoc=%d auth_alg=%u ft_over_ds=%u reassoc=%d authorized=%d ft_tk=%d fils_tk=%d)",
+		   " (added_unassoc=%d auth_alg=%u ft_over_ds=%u reassoc=%d authorized=%d ft_tk=%d fils_tk=%d, type=%d)",
 		   MAC2STR(sta->addr), sta->added_unassoc, sta->auth_alg,
 		   sta->ft_over_ds, reassoc,
 		   !!(sta->flags & WLAN_STA_AUTHORIZED),
 		   wpa_auth_sta_ft_tk_already_set(sta->wpa_sm),
-		   wpa_auth_sta_fils_tk_already_set(sta->wpa_sm));
+		   wpa_auth_sta_fils_tk_already_set(sta->wpa_sm), type);
 
 	if (!ap_sta_is_mld(hapd, sta) && !sta->added_unassoc &&
 	    (!(sta->flags & WLAN_STA_AUTHORIZED) ||
@@ -5390,7 +5391,8 @@ static int add_associated_sta(struct hostapd_data *hapd,
 			    sta->vht_opmode, sta->p2p_ie ? 1 : 0,
 			    set, mld_link_addr, mld_link_sta, eml_cap,
 			    sta->flags & WLAN_STA_EHT? sta->eht_240mhz_capab : NULL,
-			    sta->flags & WLAN_STA_EHT? sta->eht_240mhz_len : 0)) {
+			    sta->flags & WLAN_STA_EHT? sta->eht_240mhz_len : 0,
+			    type)) {
 		hostapd_logger(hapd, sta->addr,
 			       HOSTAPD_MODULE_IEEE80211, HOSTAPD_LEVEL_NOTICE,
 			       "Could not %s STA to kernel driver",
diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index 41929d3..ec72d9b 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -2590,6 +2590,7 @@ hostapd_handle_link_reconf_req(struct hostapd_data *hapd, const u8 *buf,
 	u8 dialog_token;
 	const struct ieee80211_mgmt *mgmt = (const struct ieee80211_mgmt *) buf;
 	struct link_reconf_req_list *req_list = NULL;
+	struct ml_reconf_req ml_reconf_req = {};
 	const u8 *pos = NULL;
 	int ret = -1;
 
@@ -2700,6 +2701,15 @@ skip_oci_validation:
 	if (!hostapd_validate_link_reconf_req(hapd, assoc_sta, req_list))
 		goto out;
 
+	os_memcpy(ml_reconf_req.addr, assoc_sta->addr, ETH_ALEN);
+	for (int i = 0; i < MAX_NUM_MLD_LINKS; i++) {
+		ml_reconf_req.sta_add_params[i] =
+			assoc_sta->recfg_sta_add_params[i];
+	}
+	ml_reconf_req.add_links = req_list->links_add_ok;
+	ml_reconf_req.del_links = req_list->links_del_ok;
+	ml_reconf_req.valid_links = req_list->new_valid_links;
+
 	req_list->dialog_token = dialog_token;
 	ret = hostapd_send_link_reconf_resp(hapd, assoc_sta, req_list);
 	if (ret) {
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 4690220..70d4002 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -2205,7 +2205,7 @@ int ap_sta_re_add(struct hostapd_data *hapd, struct sta_info *sta, int check_aut
 			    0, NULL, NULL, NULL, 0, NULL, 0, NULL,
 			    sta->flags, 0, 0, 0, 0,
 			    mld_link_addr, mld_link_sta, eml_cap,
-			    NULL, 0)) {
+			    NULL, 0, 0)) {
 		hostapd_logger(hapd, sta->addr,
 			       HOSTAPD_MODULE_IEEE80211,
 			       HOSTAPD_LEVEL_NOTICE,
-- 
2.34.1

